src/main/java/frc/robot/events/EventWatcherThread.java

```diff
@@ -0,0 +1,50 @@
+package frc.robot.events;
+
+import edu.wpi.first.wpilibj.command.Scheduler;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+
+/** Singleton thread for monitoring things. */
+public class EventWatcherThread extends Thread {
+
+  private ArrayList<Event> queue = new ArrayList<>();
+  private LinkedHashMap<Event, Boolean> lastStateCache = new LinkedHashMap<>();
+
+  private static EventWatcherThread thread = new EventWatcherThread();
+
+  public static EventWatcherThread getInstance() {
+    return thread;
+  }
+
+  @Override
+  public void run() {
+
+    //noinspection InfiniteLoopStatement
+    while (true) {
+      for (Event e : queue) {
+
+        // We have to store this because e.check() might change
+        boolean savedState = e.check();
+
+        // Add the event to the cache if it wasn't already there
+        if (!lastStateCache.containsKey(e)) {
+          lastStateCache.put(e, savedState);
+        }
+
+        // Add the command to the scheduler only if the state of the event changed
+        if (savedState && !lastStateCache.get(e)) {
+          Scheduler.getInstance().add(e.getCommand());
+        }
+
+        lastStateCache.replace(e, savedState);
+      }
```

Comment:
to exit this while(true) cleanly, you'd need a flag to check on each iteration. Like bIsExiting.

Reply:
yup.